File: src/main/csharp/MazeRunner.cs
namespace CodeWars
{
    // https://www.codewars.com/kata/58663693b359c4a6560001d6
    class Kata
    {
        class Position {
          private int x;
          private int y;

          public Position(int x, int y) {
            this.x = x;
            this.y = y;
          }

          public int getX() { return x; }
          public int getY() { return y; }


        public Position move(string direction) {
          switch(direction) {
              case "N":
                  return new Position(getX()-1, getY());
              case "S":
                  return new Position(getX()+1, getY());
              case "E":
                  return new Position(getX(), getY()+1);
              case "W":
                  return new Position(getX(), getY()-1);
          }

          return this;
        }
}

      class Maze {
        private int[,] maze;


        public Maze(int[,] maze) {
          this.maze = maze;
        }

        private int getDimension() { return maze.GetLength(0); }

        public Position getStartPosition() {
          for (int x = 0; x < maze.GetLength(0); x++) {
            for (int y = 0; y < maze.GetLength(1); y++) {
              if (maze[x, y] == 2) {
                return new Position(x, y);
              }
            }
          }

          return null;
        }

        public bool isOutside(Position position) {
            return position.getX() < 0 || position.getX()>= getDimension()
               || position.getY() < 0 || position.getY() >= getDimension();
        }

        public bool isEndPoint(Position position) {
          return GetPointValue(position) == 3;
        }
        public bool hitWall(Position position) {
          return GetPointValue(position) == 1;
        }

        private int GetPointValue(Position position) {
          return maze[position.getX(),position.getY()];
        }
      }

        public string mazeRunner(int[,] maze, string[] directions)
        {
          Maze m = new Maze(maze);
          Position actualPosition = m.getStartPosition();

          foreach(string direction in directions) {
            actualPosition = actualPosition.move(direction);
            if(m.isOutside(actualPosition) || m.hitWall(actualPosition)) {
              return "Dead";
            }
            if (m.isEndPoint(actualPosition)) {
              return "Finish";
            }
          }

          return "Lost";
        }
      }
}

File: src/main/c/MazeRunner.c
#include <stddef.h>
#include <string.h>

int *findStartPosition(const int *maze, size_t dimension) {
  int *actual = maze;
  int *max = maze + dimension*dimension;
  for(int *actual=maze; actual <= max; actual++) {
    if (*actual == 2) {
      return actual;
    }
  }
  return -1;
}

int *move(int *position, size_t dimension, char direction) {
  switch(direction) {
          case 'N':
            position -= dimension;
            break;
          case 'S':
            position += dimension;
            break;
          case 'E':
            position++;
            break;
          case 'W':
            position--;
            break;
      }
  
  return position;
}

int isOutside(int dimension, int *min, int* max, int* actual, int *before) {
  int diff1 = before-min;
  int diff2 = actual-min;
  
  return 
    // leave maze south or north
    actual > max || actual < min ||
    // leave maze west
    diff1%dimension == 0 && diff2%dimension == dimension-1 ||
    // leave maze east
    diff2%dimension == 0 && diff1%dimension == dimension-1
        ;
}

char *maze_runner(const int *maze, size_t n, const char *directions) {

    int *max = maze + n*n;
    int *actual = findStartPosition(maze, n);
    int moves = strlen(directions);

  for(int i = 0; i < moves; i++, directions++) {
      int *before = actual;
      actual = move(actual, n, *directions);
    
      if (isOutside(n, maze, max, actual, before)
         || *actual == 1) {
        return "Dead";
      }
      if (*actual == 3) {
        return "Finish";
      }
    }
  
    //  <----  hajime!
    return "Lost";
}
